Fix #4689: Add regex check to prohibit post() and postDelayed()

kkmurerwa opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fix #4689: When merged, this PR will;
- Add regex check to prohibit the usage of post() and postDelayed()
- Add tests for the regex check added.
- Add file exemptions for files that don't meet the new rules.

**The commits pushed on 4th April 2023 will**;
- Remove tests from exemptions list of the regex check.
- Add `RegexPatternValidationCheckTest.kt` to exemptions since it is matched correctly against the regex. This is an undesired side effect or removing tests from exemptions.
- Modify the language of the error message to align with Oppia guidelines.
- Remove the `SpotlightFragment.kt` from exemptions since it contains no instance of post() or postDelayed()
- Update the regex to take care of cases with curly braces such as `post {}` instead of only matching `post() `and `postDelayed()`.
- Cater for any spaces between the post/postDelayed and the bracket used i.e. `post ()`, `postDelayed ()` and `post {}`.
- Add another test to verify check for usages of post with curly braces.
- Add an extra test in the test methods to ensure that the correct error message is being shown.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

kkmurerwa committed:
chore: This it will;
- Add regex check to prohibit the usage of post() and postDelayed()
- Add tests for the regex check added

kkmurerwa committed:
chore: Add more files with instances of post() and postDelayed() to exemptions

kkmurerwa committed:
Merge branch 'develop' into regex-to-prohibit-post-and-post-delayed

kkmurerwa commented:
Hey @BenHenning. I did some housekeeping here and merged the latest changes from develop. All checks were passing before the merge and should even after the merge. The failing checks just need a rerun.

BenHenning committed:
Merge branch 'develop' into regex-to-prohibit-post-and-post-delayed

BenHenning commented on scripts/assets/file_content_validation_checks.textproto:
```diff
@@ -408,3 +408,13 @@ file_content_checks {
+file_content_checks {
+  file_path_regex: ".+?.kt"
+  prohibited_content_regex: ".+?\\.(post|postDelayed)\\(.+?"
+  failure_message: "Using post() and postDelayed() method is prohibited as it can lead to NullPointerException errors. Use doOnPreDraw or doOnLayout instead. Checkout this resource for a better understanding of the issue: https://betterprogramming.pub/stop-using-post-postdelayed-in-your-android-views-9d1c8eeaadf2"
```
```suggestion
  failure_message: "Prefer avoiding post() and postDelayed() method as they can can lead to subtle and difficult-to-debug crashes. Prefer using LifecycleSafeTimerFactory for most cases when an operation needs to run at a future time. For cases when state needs to be synchronized with a view, use doOnPreDraw or doOnLayout instead. For more context on the underlying issue, see: https://betterprogramming.pub/stop-using-post-postdelayed-in-your-android-views-9d1c8eeaadf2."
```

Softened up the language a bit, and added to the recommendation per Oppia best practices.

BenHenning replied:
This has been applied in the latest changes.

BenHenning commented on scripts/assets/file_content_validation_checks.textproto:
```diff
@@ -408,3 +408,13 @@ file_content_checks {
+  exempted_file_name: "app/src/main/java/org/oppia/android/app/topic/info/TopicInfoFragmentPresenter.kt"
+  exempted_file_name: "app/src/main/java/org/oppia/android/app/drawer/NavigationDrawerFragmentPresenter.kt"
+  exempted_file_name: "app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt"
+  exempted_file_patterns: ".+?Test\\.kt"
```
Do any tests actually use post/postDelayed? I don't think we want to exempt this there, either, as tests probably shouldn't be using these functions either.

BenHenning replied:
Please make sure to respond to this @kkmurerwa.

kkmurerwa replied:
I have removed the exemption for all tests and only added the `RegexPatternValidationCheckTest` since it contains an instance of post/postDelayed in the prohibited content string.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:
```diff
@@ -2289,6 +2289,40 @@ class RegexPatternValidationCheckTest {
+      runScript()
+    }
+
+    assertThat(exception).hasMessageThat().contains(REGEX_CHECK_FAILED_OUTPUT_INDICATOR)
```
Please make sure the correct error is being checked in this & the test above, similar to how we do for all other tests in this suite.

BenHenning replied:
This appears to be resolved now.

BenHenning reviewed (changes requested):
Thanks @kkmurerwa! Had some follow-up thoughts, PTAL.

kkmurerwa committed:
Merge branch 'develop' into regex-to-prohibit-post-and-post-delayed

kkmurerwa committed:
feat: Remove Test files from exemptions and add new test to check whether error message displayed is correct

kkmurerwa committed:
fix: Fix lint check issues

kkmurerwa commented:
PTAL @kkmurerwa

kkmurerwa committed:
feat: Modified regex to cater for cases where post is used as post {} instead of post()

kkmurerwa committed:
test: Add test for new curly braces condition for post {}

kkmurerwa commented:
Hey @BenHenning. I updated this PR with some new changes. They are detailed in a second section of the PR comment labeled `Latest commit will`. I have also addressed all issues mentioned and I hope it is now good to go.

kkmurerwa commented:
PTAL @BenHenning

kkmurerwa committed:
This commit will;
- Refactor the prohibited regex to cover false positives in some files
- Remove the wrongly exempted SpotlightFragment.kt file

kkmurerwa commented:
Hey @BenHenning. A few checks failed after the last commit but it shouldn't be a problem. A quick rerun should fix it.

MohitGupta121 commented:
@kkmurerwa Done, I re-run the CI checks. Thanks.

kkmurerwa commented:
Thanks @MohitGupta121

BenHenning commented on scripts/assets/file_content_validation_checks.textproto:
```diff
@@ -410,11 +410,12 @@ file_content_checks {
-  failure_message: "Using post() and postDelayed() method is prohibited as it can lead to NullPointerException errors. Use doOnPreDraw or doOnLayout instead. Checkout this resource for a better understanding of the issue: https://betterprogramming.pub/stop-using-post-postdelayed-in-your-android-views-9d1c8eeaadf2"
-  exempted_file_name: "app/src/main/java/org/oppia/android/app/spotlight/SpotlightFragment.kt"
+  prohibited_content_regex: ".+?\\.(post|postDelayed)[\\s]*(\\(|\\{).*?"
+  failure_message: "Prefer avoiding post() and postDelayed() method as they can can lead to subtle and difficult-to-debug crashes. Prefer using LifecycleSafeTimerFactory for most cases when an operation needs to run at a future time. For cases when state needs to be synchronized with a view, use doOnPreDraw or doOnLayout instead. For more context on the underlying issue, see: https://betterprogra ...
```
```suggestion
  failure_message: "Prefer avoiding post() and postDelayed() methods as they can can lead to subtle and difficult-to-debug crashes. Prefer using LifecycleSafeTimerFactory for most cases when an operation needs to run at a future time. For cases when state needs to be synchronized with a view, use doOnPreDraw or doOnLayout instead. For more context on the underlying issue, see: https://betterprogramming.pub/stop-using-post-postdelayed-in-your-android-views-9d1c8eeaadf2."
```

kkmurerwa replied:
I have updated the failure message with the suggestion. This has been fixed.

kkmurerwa replied:
I have modified the failure message to fix the grammatical error identified here. I have also updated the text on the `doesNotUsePostOrPostDelayed` variable to ensure tests for the correct error still pass.

BenHenning commented on scripts/assets/file_content_validation_checks.textproto:
```diff
@@ -410,11 +410,12 @@ file_content_checks {
   exempted_file_name: "app/src/main/java/org/oppia/android/app/drawer/NavigationDrawerFragmentPresenter.kt"
   exempted_file_name: "app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt"
-  exempted_file_patterns: ".+?Test\\.kt"
+  exempted_file_name: "utility/src/main/java/org/oppia/android/util/parser/image/UrlImageParser.kt"
+  exempted_file_name: "app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt"
+  exempted_file_name: "scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt"
```
Please sort the exempted_file_names in lexicographical order.

kkmurerwa replied:
The latest commit has re-organized the exempted files in lexicographic order.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:
```diff
@@ -168,6 +168,13 @@ class RegexPatternValidationCheckTest {
     "Use AnalyticsStartupListener to retrieve an instance of WorkManager rather than fetching one" +
       " using getInstance (as the latter may create a WorkManager if one isn't already present, " +
       "and the application may intend to disable it)."
+  private val doesNotUsePostorPostDelayed =
```
```suggestion
  private val doesNotUsePostOrPostDelayed =
```

kkmurerwa replied:
This has been updated with the latest commit alongside any uses of the variable in the subsequent tests.

BenHenning reviewed (changes requested):
Thanks @kkmurerwa. Took another pass--please make sure to respond to all comments before sending PRs back into review.

kkmurerwa committed:
This commit ;
- Rearrange the exempted files n lexicographical order.
- Rename the doesNotUsePostorPostDelayed to doesNotUsePostOrPostDelayed
- Modify the failure message to fix grammatical errors

kkmurerwa committed:
Merge branch 'develop' into regex-to-prohibit-post-and-post-delayed

kkmurerwa commented:
@BenHenning I have responded to all unresolved comments and fixed the issues highlighted. I believe it is ready for another review.

kkmurerwa commented:
PTAL @BenHenning

BenHenning reviewed (approved):
Thanks @kkmurerwa! This LGTM!
